contracts/DefaultStrategy.sol

```diff
@@ -522,248 +436,288 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
+        if (group == head) {
+            uint256 divisionResidue = totalStCeloInDefaultStrategy -
+                (expectedStCelo * activeGroups.getNumElements());
+            expectedStCelo += divisionResidue;
+        }
 
-            subtractFromStrategyTotalStCeloVotes(
-                deprecatedGroupsWithdrawn[i],
-                currentStCeloInStrategy
-            );
+        actualStCelo = stCELOInGroup[group];
+    }
 
-            if (remainingWithdrawal == 0) {
-                break;
-            }
+    /**
+     * @notice Adds/substracts value to totals of group and
+     * total stCELO in default strategy.
+     * @param group The validator group that we are updating.
+     * @param stCeloAmount The amount of stCELO.
+     * @param add Whether to add or substract.
+     */
+    function updateGroupStCelo(
+        address group,
+        uint256 stCeloAmount,
+        bool add
+    ) internal {
+        if (add) {
+            stCELOInGroup[group] += stCeloAmount;
+            totalStCeloInDefaultStrategy += stCeloAmount;
+        } else {
+            stCELOInGroup[group] -= stCeloAmount;
+            totalStCeloInDefaultStrategy -= stCeloAmount;
         }
+    }
 
-        return (
-            deprecatedGroupsWithdrawn,
-            deprecatedWithdrawalsPerGroup,
-            numberDeprecatedGroupsWithdrawn,
-            remainingWithdrawal
-        );
+    function getLesserAndGreaterOfActiveGroups(
+        address originalKey,
+        uint256 newValue,
+        bool withdrawal
+    ) internal view returns (address previous, address next) {
+        if (withdrawal) {
+            (previous, next) = getLesserAndGreaterOfActiveGroupsWithdrawal(originalKey, newValue);
+        } else {
+            (previous, next) = getLesserAndGreaterOfActiveGroupsDeposit(originalKey, newValue);
+        }
     }
 
     /**
-     * @notice Calculates how votes should be withdrawn from each active group.
-     * @param withdrawal The number of votes that need to be withdrawn.
-     * @return The array of group addresses that should be withdrawn from.
-     * @return The amount of votes to withdraw from the respective group in the
-     * array of groups withdrawn from.
+     * @notice Marks a group as deprecated.
+     * @param group The group to deprecate.
      */
-    function getActiveGroupWithdrawalDistribution(uint256 withdrawal)
-        internal
-        returns (address[] memory, uint256[] memory)
-    {
-        uint256 numberGroups = activeGroups.length();
-
-        if (withdrawal == 0 || numberGroups == 0) {
-            address[] memory noGroups = new address[](0);
-            uint256[] memory noWithdrawals = new uint256[](0);
-            return (noGroups, noWithdrawals);
+    function _deprecateGroup(address group) private {
+        if (!activeGroups.contains(group)) {
+            revert GroupNotActive(group);
+        }
+        activeGroups.remove(group);
+        unsortedGroups.remove(group);
```

Comment:
Won't this fail if group was not unsorted? Which could be the case for a group we want to deactivate, if I understand correctly.

Reply:
According to documentation it will just return false in such case

https://docs.openzeppelin.com/contracts/3.x/api/utils#EnumerableSet-remove-struct-EnumerableSet-AddressSet-address-

Reply:
Ah right, was thinking of SortedLinkedList, but of course `unsortedGroups` is not that data structure.